File: two_sum/two_sum.py
def twoSum(nums, target):
    """
    :type nums: List[int]
    :type target: int
    :rtype: List[int]
    """
    numToIndex = {}
    for index, num in enumerate(nums):
      complement = target - num
      if complement in numToIndex:
        return [numToIndex[complement], index]
      numToIndex[num] = index
    return []

print(twoSum([2,5,5,11], 10)) # [ 1, 2 ]
print(twoSum([3,2,4], 6))     # [ 1, 2 ]
print(twoSum([3,3], 6))       # [ 0, 1 ]

File: two_sum/TwoSum.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

class TwoSum {
  public static int[] twoSum(int[] nums, int target) {
    Map<Integer, Integer> numToIndex = new HashMap<Integer, Integer>(nums.length);

    for (int i = 0; i < nums.length; i++){
      int num = nums[i];
      int complement = target - num;

      if (numToIndex.containsKey(complement))
        return new int[] {numToIndex.get(complement), i};

      numToIndex.put(num, i);
    }
    return new int[0];
  }

  public static void main(String[] args){
    // int[] nums = {2,5,5,11}; int target = 10; // [ 1, 2 ]
    int[] nums = {3,2,4}; int target = 6;     // [ 1, 2 ]
    // int[] nums = {3,3}; int target = 6;       // [ 0, 1 ]

    int[] resPair = twoSum(nums, target);

    System.out.println(Arrays.toString(resPair));
  }
}
